[Mission5/신동호] Project_Notion_VanillaJs 과제

khakhiD opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->
## 🌎 배포 링크
https://vanilla-js-notion-clone-project.vercel.app/

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
바닐라 JS만을 이용해 노션을 클로닝합니다.
기본적인 레이아웃은 노션과 같으며, 스타일링, 컬러값 등은 원하는대로 커스텀합니다.
글 단위를 Document라고 합니다. Document는 Document 여러개를 포함할 수 있습니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
### 기본 요구사항
- [x] 화면 좌측에 Root Documents를 불러오는 API를 통해 루트 Documents를 렌더링합니다.
- [x] Root Document를 클릭하면 오른쪽 편집기 영역에 해당 Document의 Content를 렌더링합니다.
- [x] 해당 Root Document에 하위 Document가 있는 경우, 해당 Document 아래에 트리 형태로 렌더링 합니다.
- [x] Document Tree에서 각 Document 우측에는 + 버튼이 있습니다. 해당 버튼을 클릭하면, 클릭한 Document의 하위 Document로 새 Document를 생성하고 편집화면으로 넘깁니다.
- [x] 편집기에는 기본적으로 저장 버튼이 없습니다. Document Save API를 이용해 지속적으로 서버에 저장되도록 합니다.
- [x] History API를 이용해 SPA 형태로 만듭니다.
- [x] 루트 URL 접속 시엔 별다른 편집기 선택이 안 된 상태입니다.
- [x] /documents/{documentId} 로 접속시, 해당 Document 의 content를 불러와 편집기에 로딩합니다.

### 보너스 요구사항
- [ ] 기본적으로 편집기는 textarea 기반으로 단순한 텍스트 편집기로 시작하되, 여력이 되면 div와 contentEditable을 조합해서 좀 더 Rich한 에디터를 만들어봅니다.
- [x] 편집기 최하단에는 현재 편집 중인 Document의 하위 Document 링크를 렌더링하도록 추가합니다.
- [ ] 편집기 내에서 다른 Document name을 적은 경우, 자동으로 해당 Document의 편집 페이지로 이동하는 링크를 거는 기능을 추가합니다.

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 기본 요구사항에 충실히 준수하려고 했습니다.
- 컴포넌트 구조를 깊게 할 수록 머리가 아파와서 최대한 단순히 하려고 했습니다.

## ✅ 리팩토링 반영 사항
- 무의미한 eslint 설정 및 babel 프로젝트에서 삭제
  - jest 코드 작성을 못해서 eslint 관련 파일도 삭제했습니다. 다음 과제에서는 테스트 코드를 작성해보도록 하겠습니다.
- 컴포넌트 내부에 사용되는 이벤트, 클래스 명 상수 처리
- 방어 코드 추가
- 조건문 표현 리팩토링 (옵셔널 체이닝 사용 등)
- Editor 컴포넌트: querySelector를 한 번만 사용하기 위해서 요소 참조를 변수 방식으로 변경
- deprecated 코드인 e.keyCode 대신 e.key로 리팩토링
- debounce 함수 따로 구현 후 onEdit 함수에 적용
- docsPage에서 document 클릭 시 두 번 네트워크 요청되는 부분 수정
- 서버에 저장 중임을 알리는 로딩 아이콘 기능 추가
  - Loading 컴포넌트 추가
  - Editor 컴포넌트에서 keyup 이벤트 발생 시 Loading 렌더링
  - Loading 컴포넌트 관련 스타일 추가
- 하위 컴포넌트들의 state 구조 수정
- fontawesome 설치형 방식으로 변경 (index.html에서 cdn 관련 링크 및 스크립트 태그 삭제)
- 패키지 설정 및 vite 설정 추가

oaoong committed:
:tada: 기초 환경 설정

khakhiD committed:
remove: hi.js hi.test.js

khakhiD committed:
remove: hi.test.js

khakhiD committed:
remove: index.js

khakhiD committed:
feat: 추가 문서 리스트 바

khakhiD committed:
feat: 추가 Router 기능

khakhiD committed:
feat: 추가 API 요청기능

khakhiD committed:
feat: 추가 문서 편집 페이지, 편집 기능

khakhiD committed:
chore: settings

khakhiD committed:
feat: normalize, style 적용

khakhiD committed:
chore: vercel.json

khakhiD committed:
chore: vercel.json

khakhiD committed:
style: nowrap -> wrap

khakhiD committed:
feat: 루트문서 삭제되버림

khakhiD committed:
style: 주먹구구식 margin-right

khakhiD committed:
feat: escapeHTML 기능 추가 및 적용

oaoong commented on src/components/DocsList/DocsList.js:
```diff
@@ -0,0 +1,82 @@
+      if (classList.contains("create")) {
+        const newDocId = await onCreate(id)
+        push(`/documents/${newDocId}`)
+      } else if (classList.contains("remove")) {
```
```suggestion
      } else if (classList.contains("remove") || e.target.closest(".remove")) {
```
remove 버튼을 누르다가 가끔씩 작동을 안해서 디버깅해보니 svg 태그 안의 path가 클릭 target으로 지정되는 경우는 인식을 못하더라고요, 위와같이 추가적인 체크를 해주니 잘 작동하는 것 같습니다!

khakhiD replied:
와 저는 계속 contains, closest, tagName ... 등등의 조건 써가면서 그나마 잘 동작하는 걸 넣었다고 생각했거든요.
그냥 `||` 넣어주면 되는건데 지금 보니까 왜 계속 조건을 하나로 넣었는지 의문이네요.
제가 오직 하나의 조건으로 엄청 깔끔하게 보이도록 작성하고 싶었나 봅니다..
감사합니다~

dlwl98 commented on src/utils/escapeHTML.js:
XSS 공격 방어코드 좋네요!

dlwl98 commented on src/components/App.js:
```diff
@@ -0,0 +1,54 @@
+export default function App({ $target }) {
+  let timer = null
+
+  const onEdit = async ({ id, title, content }) => {
+    if (timer !== null) {
+      clearTimeout(timer)
+    }
+    timer = setTimeout(async () => {
+      await request(`/documents/${id}`, {
+        method: "PUT",
+        body: JSON.stringify({
+          title,
+          content,
+        }),
+      })
+      await docsPage.render()
+      // editPage.render()
+    }, 1000)
+  }
```
디바운스 로직이 여기서만 쓰여서 분리할필요가 없지만 분리할 경우 테스트코드 작성 등 부수적인 득이 있습니다.
또한 의식적으로 도메인로직을 덜어내는 연습을 하는 것이 좋다고 생각해요. 저는 이런게 어렵더라구요.

khakhiD replied:
아직 코드를 작성하면서 도메인 로직인 것과 아닌 것을 생각하면서 작성하는 습관이 부족한 것 같습니다.
리팩토링 기간에 의식해서 수정해보도록 하겠습니다. 감사합니다.

yrnana replied:
디바운스 함수를 만들어서 사용해주면 좋을 것 같습니다:-)

dlwl98 commented on src/components/DocsEdit/EditPage.js:
```diff
@@ -0,0 +1,57 @@
+    initialState: [],
+  })
+
+  this.setState = async (nextState) => {
+    if (this.state.id !== nextState.id) {
+      this.state = nextState
+      await fetchDocument()
+      return
+    }
+    this.state = nextState
+
+    editor.setState(
+      this.state || {
+        title: "",
+        content: "",
+      },
+    )
+
+    editorFooter.setState(this.state.documents)
+    this.render()
+  }
```
setState라는 메서드명과는 관련없는 로직들이 들어있어요.
setState라는 이름을 그대로 둔다면 관련없는 로직들을 덜어내는것이 좋을 것 같고,
메서드명을 다르게 바꿔도 좋지않나 싶어요.
제 생각은 함수는 이름만으로 그 역할을 예상할 수 있어야 된다고 생각해요.
이름을 보고 예상한 것과 로직이 다르면 이해하기가 어려워지니까요.

khakhiD replied:
저도 코드를 작성하면서 헷갈리고 갇힌 부분이었는데, 짚어주셔서 감사합니다.

dlwl98 commented on src/components/DocsEdit/Editor.js:
```diff
@@ -0,0 +1,59 @@
+
+  this.state = initialState
+
+  this.setState = (nextState) => {
+    this.state = nextState
+    $editor.querySelector("[name=title]").value = this.state.title
+    $editor.querySelector("[name=content]").value = this.state.content
+    this.render()
+  }
```
여기도 동일합니다. 17, 18번 줄은 this.render 에 더 어울리는 것 같아요.

hyoribogo commented on src/components/DocsEdit/Editor.js:
```diff
@@ -0,0 +1,59 @@
+  })
+
+  $editor.addEventListener("keyup", (e) => {
+    if (e.target.tagName === "INPUT" && e.keyCode === 13) {
```
13이 어떤 의미인지 따로 상수 처리를 해주면 더 이해하기 편할 것 같아요!

khakhiD replied:
맞네요. 상수로 처리했으면 키코드 13이 엔터인 걸 몰라도 바로 이해할 수 있겠네요.
감사합니다.

hyoribogo commented on src/components/App.js:
```diff
@@ -0,0 +1,54 @@
+    if (pathname === "/") {
+      docsPage.render()
+      return
+    } else if (pathname.startsWith("/documents/")) {
```
오오 startsWith라는 것 처음 알아갑니다

Jeong-Taeho replied:
저도 처음보는데 배워갑니다!!

hyoribogo commented on src/components/App.js:
```diff
@@ -0,0 +1,54 @@
+      docsPage.render()
+      return
+    } else if (pathname.startsWith("/documents/")) {
+      const [, , documentId] = pathname.split("/")
+      await editPage.setState({
+        id: documentId,
```
```js
const [, , id] = pathname.split("/")
await editPage.setState({
  id,
```

이렇게 써도 좋을 거 같아요!

hyoribogo commented on src/components/DocsEdit/EditorFooter.js:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  this.render = () => {
+    if (this.state.length > 0) {
```
`if (this.state.length)`
로 표현해도 좋을 듯 합니당

dlwl98 commented on src/components/DocsEdit/Editor.js:
```diff
@@ -0,0 +1,59 @@
+  })
+
+  $editor.addEventListener("keyup", (e) => {
+    if (e.target.tagName === "INPUT" && e.keyCode === 13) {
+      const $textarea = document.querySelector("textarea[name=content]")
+      $textarea.focus()
+    }
```
keycode 13 이 엔터군요. 배워가요~

khakhiD replied:
다음에는 상수로 처리해서 이해하기 쉽도록 해보겠습니다~

hyoribogo commented on src/utils/escapeHTML.js:
오오 변환해주는 거 신기하네요

khakhiD replied:
2번 요청되는 이유 찾아서 수정중입니당...

dlwl98 commented on src/components/DocsEdit/Editor.js:
```diff
@@ -0,0 +1,59 @@
+export default function Editor({
+  $target,
+  initialState = {
+    title: "",
+    content: "",
+  },
```
이런 부분이 사소하지만 state의 프로터티를 바로 알 수 있어서 좋네요.

JIY00N2 replied:
저도 동의합니다!

hyoribogo reviewed (commented):
동호님 잘 봤습니다.
파일 하나에 다 있으니 사실상 제일 보기 편했네요..... 😋

![image](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/assets/97094709/fa398705-6cbc-4554-a616-9f29577c4582)

새로고침을 하면 이런 요청을 하던데 혹시 뭔지 아시나요 ?
그리고 이벤트 발생 했을 때 요청을 2번씩 보내더라고요 !!

코드에서 찾아보려 했지만 못 찾았네요 .. 죄송합니닷
2번씩 요청 보내지는 거 수정하면 좋을 거 같습니다 !

dlwl98 commented on src/components/App.js:
```diff
@@ -0,0 +1,54 @@
+    if (pathname === "/") {
+      docsPage.render()
+      return
+    } else if (pathname.startsWith("/documents/")) {
```
startsWith 하나 알아가요~

dlwl98 commented on src/components/DocsEdit/EditPage.js:
```diff
@@ -0,0 +1,57 @@
+    $target.appendChild($page)
+  }
+
+  const fetchDocument = async () => {
+    const { id } = this.state
+    if (id !== "new") {
+      const doc = await request(`/documents/${id}`, { method: "GET" })
+      this.setState(doc)
+    } else {
+      this.setState()
+    }
+  }
```
저는 fetch라는 네이밍을 가지는 함수가 setState도 하는 것이 이상하다고 생각했는데
다른분들 코드에서 자주 보이네요.. 컴포넌트 입장에서의 fetch를 생각해보면 그럴듯하기도 하고?

khakhiD replied:
아마 강의에서 이런 방식으로 동작시키기 때문에 많이 보시지 않았나 싶습니다.
네이밍에 대한 고민 없이 강의를 그대로 따라 작성해서 그렇습니다.

dlwl98 reviewed (commented):
코드가 일관성있고 보기편했습니다. 강의에 굉장히 충실한 코드네요!
갈아엎기 전 코드도 궁금하네요

Jeong-Taeho commented on src/components/DocsEdit/Editor.js:
```diff
@@ -0,0 +1,59 @@
+  })
+
+  $editor.addEventListener("keyup", (e) => {
+    if (e.target.tagName === "INPUT" && e.keyCode === 13) {
```
https://way-be-developer.tistory.com/244

`e.keyCode` 가 `Deprecated` 되어 동작은 되지만 권장하지 않는다고 하더라구요. 대체 방법이 있는 것 같아 찾아보았습니다!

dlwl98 replied:
오! 좋은 정보 감사합니다. 😀

khakhiD replied:
대체 방법을 찾아주시다니 너무 감사합니다!

yrnana replied:
e.key를 쓰는게 더 좋을 것 같습니다:-)

Jeong-Taeho reviewed (commented):
동호님 과제하느라 수고하셨습니다! 깔끔하게 잘 정리되어 있어서 이해하기 편했던 것 같아요! 아는게 많이 없어서 리뷰를 제대로 못한 것 같네요😂 저도 효리님 리뷰보고 네트워크가 2번씩 계속 호출되는 이유를 찾아보려 했는데 어렵네요...

JIY00N2 commented on index.html:
```diff
@@ -0,0 +1,15 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Notion Cloning</title>
+    <script src="https://kit.fontawesome.com/1e40242dee.js" crossorigin="anonymous"></script>
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.1/normalize.min.css">
```
제가 지난번에 멘토님한테 받은 피드백인데 동호님도 참고 하셨으면 좋겠습니다!
[링크](https://github.com/prgrms-fe-devcourse/FEDC4-3_VanillaJS_1/pull/31#discussion_r1243978883)

JIY00N2 commented on src/utils/router.js:
```diff
@@ -0,0 +1,22 @@
+
+export const push = (nextUrl) => {
+  window.dispatchEvent(
+    new CustomEvent("route-change", {
```
여기 부분도 상수를 넣어주면 될 것 같아요!

khakhiD replied:
혹시 한 번만 사용되는 곳에도 상수를 적용해야 하는지 궁금해요!

JIY00N2 reviewed (commented):
동호님 코드가 강의와 유사해서 이해하기 편했습니다!
되게 디테일한 부분도 신경 쓰신 게 느껴졌습니다.
다른 팀원분들이 리뷰를 많이 해주셔서 리뷰가 적은 점 죄송합니다😥
노션 프로젝트 과제 하느라 고생 많으셨습니다!

yrnana commented on .eslintrc:
```diff
@@ -0,0 +1,13 @@
+{
+  "extends": ["eslint:recommended", "plugin:jest/recommended"],
+  "plugins": ["jest"],
```
테스트코드가 없어서 살짝 무의미한 린트 설정이 된 것 같아용

yrnana replied:
prettier와 함께 작성할때는 eslint-config-prettier를 같이 설정해주면 좋습니다
https://yrnana.dev/post/2021-03-21-prettier-eslint/

yrnana commented on src/components/DocsEdit/Editor.js:
```diff
@@ -0,0 +1,59 @@
+
+  $editor.addEventListener("keyup", (e) => {
+    if (e.target.tagName === "INPUT" && e.keyCode === 13) {
+      const $textarea = document.querySelector("textarea[name=content]")
```
이 컴포넌트의 render는 최초 한번만 발생하는 것 같은데 select를 미리 저장해놓고 사용해도 좋을 것 같네용

khakhiD replied:
[메모] render가 더 되지 않을 때에는 element를 변수에 저장해두고 querySelector를 더 불러오지 않게 하기...
감사합니다!

yrnana commented on src/components/DocsEdit/EditorFooter.js:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  $footer.addEventListener("click", (e) => {
+    const $title = e.target.closest(".sub-docs")
```
sub-doc을 상수화 하고 지정된 컴포넌트와 여기에서 사용해도 좋을 것 같아용

khakhiD replied:
[메모] class 명으로 요소를 불러올 때와 같은 경우에는 상수화해서 사용하기...
감사합니다 🥰

yrnana commented on src/components/DocsList/DocsList.js:
```diff
@@ -0,0 +1,82 @@
+  const renderDocuments = (documents) => {
+    return documents
+      .map((doc) => {
+        const hasChildDocuments = doc.documents && doc.documents.length > 0
```
`!!doc.documents?.length`
로 작성도 가능합니다 (나노팁)

khakhiD committed:
refactor: 이벤트명 상수 처리

khakhiD committed:
chore: eslint, babel 제거

khakhiD committed:
chore: 패키지 변경 및 vite 설정

khakhiD committed:
refactor: fontawesome 설치형 방식으로 변경

khakhiD committed:
chore: package management

khakhiD committed:
refactor: Editor 컴포넌트 수정
- querySelector를 한 번만 사용하기 위해서 element 참조를 변수 방식으로 변경
- setState 함수의 value 수정 부분을 render 함수 내부로 위치 변경
- keyup 이벤트에서 e.keyCode를 e.key로 변경
- input 이벤트와 key 이름을 상수화

khakhiD committed:
refactor: if문 수정
- (this.state.length > 0) => (this.state.length)

khakhiD committed:
refactor: 네트워크 요청 두 번 되는 부분 수정

khakhiD committed:
feat: debounce 유틸 함수 추가

khakhiD committed:
refactor: 클래스명 상수화

khakhiD committed:
refactor: 클래스명 상수화

khakhiD committed:
refactor: 방어 코드 추가

khakhiD committed:
feat: 로딩 아이콘 기능 추가

khakhiD committed:
refactor: 디바운스 함수 적용

khakhiD committed:
refactor: state 구조 변경

khakhiD committed:
style: Loading 아이콘 관련 스타일링 추가

khakhiD committed:
refactor: hasChildDocuments 변수에 옵셔널 체이닝 문법 적용
